Implement memoize

greg-w101 opened the pull request.

greg-w101 committed:
implement memoize

greg-w101 committed:
prettier fix

greg-w101 committed:
update readme with better example

carlansley commented on .github/workflows/ci.yml:
```diff
@@ -24,29 +21,23 @@ jobs:
         run: npm run ci:style
       - name: Perform Linting
         run: npm run ci:lint
-      - name: Run Tests
```
we should be running tests

carlansley commented on .github/workflows/ci.yml:
```diff
@@ -58,19 +49,13 @@ jobs:
         run: npm run ci:style
       - name: Perform Linting
         run: npm run ci:lint
-      - name: Run Tests
```
we should be running tests

carlansley commented on LICENSE.txt:
```diff
@@ -0,0 +1,22 @@
+
+The MIT License
+
+Copyright (c) 2021-2023 Check Digit, LLC
```
this is a new repo, should just be 2023

carlansley commented on README.md:
```diff
@@ -1,36 +1,25 @@
+# Memoize
 
-The Check Digit < Insert Name > library is a library for Check Digit services to deal with [ short description about what the library does]
+Copyright (c) 2021-2023 [Check Digit, LLC](https://checkdigit.com)
```
2023

carlansley commented on README.md:
```diff
@@ -1,36 +1,25 @@
+Copyright (c) 2021-2023 [Check Digit, LLC](https://checkdigit.com)
 
-Features:
+Memoize is a light-weight utility library used by Check Digit services for memoization purposes.
```
I think we should add an explantation that its designed to memoize promises, and evict from the cache anything that rejects.

carlansley commented on package.json:
```diff
@@ -1,52 +1,75 @@
-  "main": "./dist/index.js",
+  "name": "@checkdigit/memoize",
+  "version": "1.0.0",
+  "description": "Memoize is a light-weight utility library used by Check Digit services for memoization purposes",
```
...for memoizing promises with cache eviction on reject

carlansley commented on src/index.ts:
```diff
@@ -1,3 +1,9 @@
 
-export { exampleFunction } from './example-directory';
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
```
2023

carlansley commented on src/memoize.spec.ts:
```diff
@@ -0,0 +1,112 @@
+// memoize.spec.ts
+
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
```
2023

carlansley commented on src/memoize.spec.ts:
```diff
@@ -0,0 +1,112 @@
+ */
+
+import { strict as assert } from 'node:assert';
+import memoize from './memoize';
```
I think it should import memoize from `index.ts`, for complete code coverage

carlansley commented on src/memoize.ts:
```diff
@@ -0,0 +1,38 @@
+// memoize.ts
+
+/*
+ * Copyright (c) 2021-2023 Check Digit, LLC
```
2023

carlansley commented on src/memoize.ts:
```diff
@@ -0,0 +1,38 @@
+export type Argument = string | number | boolean | null | Date | URL | Argument[] | { [key: string]: Argument };
+
+/**
+ * Memoize an async function.  The promise is cached, not the value.  This guarantees the underlying function is
```
a version of this comment probably should be in the README

greg-w101 committed:
addressing pr comments

- change copyright to 2023
- adjust read me description
- add run tests back into ci workflows
- import memoize from index.ts for test file

greg-w101 committed:
change npm run ci:test to npm test in publish.yml

greg-w101 committed:
change node engine to >=18

lukebrdn reviewed (dismissed):
lgtm

carlansley commented on README.md:
```diff
@@ -1,36 +1,28 @@
+Copyright (c) 2023 [Check Digit, LLC](https://checkdigit.com)
 
-Features:
+Memoize is a small async function used by Check Digit services for memoizing promises with cache eviction on reject.
```
remove the "used by Check Digit services"

carlansley commented on README.md:
````diff
@@ -1,36 +1,28 @@
+### Use
 
 ```
+import { memoize } from '@checkdigit/memoize';
````
I don't think this code would work, there's only a default import.   believe should be `import memoize from '@checkdigit/memoize';`

greg-w101 committed:
remove Check Digit specific comments and fix example in read me

+ update workflows

greg-w101 commented:
> lgtm

Luke you left a comment instead of approving haha

lukebrdn reviewed (approved):
lgtm
